File: Search.SearchService/IRequestCache.cs
namespace Search.SearchService
{
    public interface IRequestCache
    {
        bool IsCached(SearchRequest request);

        SearchResponse GetResponse(SearchRequest request);
        bool TryGetResponse(SearchRequest request, out SearchResponse response);

        void Add(SearchRequest request, SearchResponse response);
        void Remove(SearchRequest request);
    }
}

File: FSharpSearchService/IRequestCache.fs
namespace FSharpSearchService

open System.Runtime.InteropServices

type IRequestCache=
    interface
        abstract IsCache: SearchRequest->bool

        abstract GetResponse:SearchRequest->SearchResponse

        abstract TryGetResponse:request:SearchRequest * [<Out>] response:SearchResponse->bool

        abstract Add:request:SearchRequest * response:SearchResponse

        abstract Remove:SearchRequest->bool
        
    end